proposals/DOM-Parts-Imperative.md

```diff
@@ -256,156 +212,67 @@ had initially contained `mailto:` before `{email}` but we could not capture this
-
-In this approach, we group multiple `AttributePart`s together by creating an
-explicit group:
-
-```js
-const firstName = new AttributePart();
-const lastName = new AttributePart();
-const group = AttributePartGroup(element, "title");
-group.append(firstName, " ", lastName);
-```
-
-This is morally equivalent to option 1 except there is an explicit grouping
-step.
-
-- **Pros**: Nicer syntax by the virtue of individual "partial" `AttributePart`'s
-  existence at the time of grouping. Code that assigns values to `AttributePart`
-  only needs to know about `AttributePart`
-- **Cons**: More objects / complexity. `AttributePart` will have two modes.
-
-### Option 3. Introduce `AttributePartFragment`
-
-Unlike option 2, this creates `AttributePartFragment`s from `AttributePart`,
-meaning that `AttributePart` in option 3 plays the role of `AttributePartGroup`
-in option 2:
+Instead of `AttributePart` having a single string `value`, it could optionally
+take an `Array` that contains values that should be concatenated together. This
+allows updating individually parts of the attribute without needing to serialize
+the entire string.
 
 ```js
-const firstNamePartial = new AttributePartFragment();
-const lastNamePartial = new AttributePartFragment();
-const part = AttributePart(element, "title");
-part.values = [firstNamePartial, " ", lastNamePartial];
+const part = AttributePart(element, "href");
+part.value = ["mailto: ", email];
 ```
 
-- **Pros**: Nicer syntax by the virtue of individual `AttributePartFragment`'s
-  existence at the time of grouping. `AttributePart` just knows one thing to do:
-  to set the whole content attribute value.
-- **Cons**: More objects / complexity. Code that uses a template has to deal
-  with two different kinds of objects: `AttributePartFragment` and
-  `AttributePart`.
-
-### Option 4. Support arbitrary JavaScript objects
-
-One way of punting is to support arbitrary JavaScript objects as `value` that
-conform to some interface. This interface could be as simple as `toString()`, or
-could use `Symbol` to determine how to populate attributes. This would allow
-code that wanted to represent partial attributes, but would maintain the
-property that parts represent nodes or groupings of nodes.
+This API may be strengthened further to optionally take in a
+`TemplateStringsArray` and variables, which would allow the browser to determine
+which parts of the attribute were compile-time constants. This could allow using
+tagged template literals to pass static content to the browser.
 
 ```js
-class TitleAttributeValue {
-  constructor() {
-    this.firstName = "";
-    this.lastName = "";
-  }
-
-  toString() {
-    return `${this.firstName} ${this.lastName}`;
-  }
-}
-
-const part = AttributePart(element, "title");
-part.value = new TitleAttributeValue();
-part.value.firstName = "Ryosuke";
-part.value.lastName = "Niwa";
+const part = AttributePart(element, "href", {
+  template: attribute`mailto: ${0}`,
```

Comment:
I'm not sure if taking a template strings array here gives us any security benefits. The static and dynamic values will just be concatted together into a single string before they're set as an attribute.

Hm, on the other hand, if native elements got information about the static and dynamic values, that might give future content security policies the ability to set constraints that some parts of attribute strings must be constant. The `safevalues` library does something like that for high risk URL attributes like an iframe's `src`, where the origin must be statically determined, but the path can be dynamic.

Ok, in that case I think I see the value here.